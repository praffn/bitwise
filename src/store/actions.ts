import { createAction } from "typesafe-actions";
import uuid from "uuid";

import {
  SET_RADIX,
  SET_BIT_LENGTH,
  TOGGLE_SIGNED,
  HISTORY_ADD,
  HISTORY_REMOVE,
  HISTORY_CLEAR,
  SET_REGISTER,
  SWAP_REGISTERS,
  TOGGLE_REGISTER_BIT
} from "./constants";
import { Radix, BitLength, RegisterKey, BinaryOp } from "../types";

export const setRadix = createAction(SET_RADIX, resolve => {
  return (radix: Radix) => resolve({ radix });
});

export const setBitLength = createAction(SET_BIT_LENGTH, resolve => {
  return (bitLength: BitLength) => resolve({ bitLength });
});

export const toggleSigned = createAction(TOGGLE_SIGNED);

export const setRegister = createAction(SET_REGISTER, resolve => {
  return (key: RegisterKey, value: number) => resolve({ key, value });
});

export const toggleRegisterBit = createAction(TOGGLE_REGISTER_BIT, resolve => {
  return (key: RegisterKey, index: number) => resolve({ key, index });
});

export const swapRegisters = createAction(SWAP_REGISTERS);

export const historyAdd = createAction(HISTORY_ADD, resolve => {
  return (op: BinaryOp) =>
    resolve({
      op,
      id: uuid()
    });
});

export const historyRemove = createAction(HISTORY_REMOVE, resolve => {
  return (id: string) => resolve({ id });
});

export const historyClear = createAction(HISTORY_CLEAR);
