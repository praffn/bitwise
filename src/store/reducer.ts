import { ActionType, getType } from "typesafe-actions";
import { BitLength, HistoryItem, Radix, Registers } from "../types";
import { compute, getBitLengthMask } from "../util";
import * as actions from "./actions";

export interface State {
  readonly signed: boolean;
  readonly bitLength: BitLength;
  readonly radix: Radix;
  readonly registers: Registers;
  readonly history: ReadonlyArray<HistoryItem>;
  readonly runningTutorial: boolean;
}

export const initialState: State = {
  bitLength: 16,
  history: [],
  radix: "dec",
  registers: {
    A: 0,
    B: 0,
  },
  runningTutorial: false,
  signed: false,
};

export default (
  state: State = initialState,
  action: ActionType<typeof actions>,
) => {
  switch (action.type) {
    case getType(actions.toggleSigned):
      return { ...state, signed: !state.signed };
    case getType(actions.setBitLength):
      const mask = getBitLengthMask(action.payload.bitLength);
      return {
        ...state,
        bitLength: action.payload.bitLength,
        registers: { A: state.registers.A & mask, B: state.registers.B & mask },
      };
    case getType(actions.setRadix):
      return { ...state, radix: action.payload.radix };
    case getType(actions.swapRegisters):
      return {
        ...state,
        registers: { A: state.registers.B, B: state.registers.A },
      };
    case getType(actions.setRegister):
      return {
        ...state,
        registers: {
          ...state.registers,
          [action.payload.key]: action.payload.value,
        },
      };
    case getType(actions.toggleRegisterBit):
      const v = state.registers[action.payload.key];
      return {
        ...state,
        registers: {
          ...state.registers,
          [action.payload.key]: (v ^ (1 << action.payload.index)) >>> 0,
        },
      };
    case getType(actions.historyAdd): {
      const { A, B } = state.registers;
      const { op, id } = action.payload;
      return {
        ...state,
        history: [
          { A, B, op, id, value: compute(A, B, state.bitLength, op) },
          ...state.history,
        ].slice(0, 10),
      };
    }
    case getType(actions.historyRemove):
      return {
        ...state,
        history: state.history.filter(({ id }) => id !== action.payload.id),
      };
    case getType(actions.historyClear):
      return {
        ...state,
        history: [],
      };
    case getType(actions.setRunningTutorial):
      return {
        ...state,
        runningTutorial: action.payload.running,
      };
    default:
      return state;
  }
};
