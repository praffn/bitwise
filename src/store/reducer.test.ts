import reducer, { State, initialState } from "./reducer";
import * as actions from "./actions";

const getInitialState = (state?: Partial<State>) =>
  ({
    ...initialState,
    ...(state ? state : {})
  } as State);

describe("reducer", () => {
  describe("initialState", () => {
    it("has correct initialState", () => {
      const initialState = reducer(undefined, {} as any);
      expect(initialState).toMatchSnapshot();
    });
  });

  describe("signed", () => {
    it("toggles ON signed", () => {
      const initialState = getInitialState({ signed: false });
      expect(initialState.signed).toBeFalsy();
      const state = reducer(initialState, actions.toggleSigned());
      expect(state.signed).toBeTruthy();
    });

    it("toggles OFF signed", () => {
      const initialState = getInitialState({ signed: true });
      expect(initialState.signed).toBeTruthy();
      const state = reducer(initialState, actions.toggleSigned());
      expect(state.signed).toBeFalsy();
    });
  });

  describe("radix", () => {
    it("can set radix", () => {
      const initialState = getInitialState({ radix: "dec" });
      expect(initialState.radix).toEqual("dec");
      const state = reducer(initialState, actions.setRadix("hex"));
      expect(state.radix).toEqual("hex");
    });
  });

  describe("bitLength", () => {
    it("can set bitLength", () => {
      const initialState = getInitialState({ bitLength: 16 });
      expect(initialState.bitLength).toEqual(16);
      const state = reducer(initialState, actions.setBitLength(32));
      expect(state.bitLength).toEqual(32);
    });
  });

  describe("registers", () => {
    it("can set registers", () => {
      const initialState = getInitialState({ registers: { A: 2, B: 3 } });
      expect(initialState.registers.A).toEqual(2);
      expect(initialState.registers.B).toEqual(3);
      const state1 = reducer(initialState, actions.setRegister("A", 100));
      expect(state1.registers.A).toEqual(100);
      expect(state1.registers.B).toEqual(3);
      const state2 = reducer(state1, actions.setRegister("B", 200));
      expect(state2.registers.B).toEqual(200);
    });

    it("can swap registers", () => {
      const initialState = getInitialState({ registers: { A: 2, B: 3 } });
      const state = reducer(initialState, actions.swapRegisters());
      expect(state.registers.A).toEqual(3);
      expect(state.registers.B).toEqual(2);
    });

    it("can toggle nth bit of register", () => {
      const initialState = getInitialState({ registers: { A: 2, B: 3 } });
      const state1 = reducer(initialState, actions.toggleRegisterBit("A", 0));
      expect(state1.registers.A).toEqual(3);
      const state2 = reducer(state1, actions.toggleRegisterBit("B", 0));
      expect(state2.registers.B).toEqual(2);
    });
  });

  describe("history", () => {
    describe("adding items", () => {
      it("can add new items", () => {
        const initialState = getInitialState();
        expect(initialState.history).toHaveLength(0);
        const state = reducer(initialState, actions.historyAdd("add"));
        expect(state.history).toHaveLength(1);
      });

      it("new items are added at the front", () => {
        const initialState = getInitialState({
          history: [{ A: 0, B: 0, id: "initial", value: 0, op: "add" }]
        });
        expect(initialState.history[0].id).toEqual("initial");
        const state = reducer(initialState, actions.historyAdd("add"));
        expect(state.history).toHaveLength(2);
        expect(state.history[0].id !== "initial").toBeTruthy();
      });

      it("caps at 10 items", () => {
        const initialState = getInitialState({
          history: new Array(10).fill({
            A: 0,
            B: 0,
            id: "",
            value: 0,
            op: "add"
          })
        });
        expect(initialState.history).toHaveLength(10);
        const state = reducer(initialState, actions.historyAdd("add"));
        expect(state.history).toHaveLength(10);
        expect(state.history[0].id !== "").toBeTruthy();
      });
    });

    describe("removing items", () => {
      it("removes item given existing id", () => {
        const initialState = getInitialState({
          history: [{ A: 0, B: 0, id: "initial", value: 0, op: "add" }]
        });
        expect(initialState.history).toHaveLength(1);
        const state = reducer(initialState, actions.historyRemove("initial"));
        expect(state.history).toHaveLength(0);
      });

      it("does nothing if id does not exists", () => {
        const initialState = getInitialState({
          history: [{ A: 0, B: 0, id: "initial", value: 0, op: "add" }]
        });
        expect(initialState.history).toHaveLength(1);
        const state = reducer(
          initialState,
          actions.historyRemove("doesnt exists")
        );
        expect(state.history).toHaveLength(1);
      });
    });

    describe("clearing", () => {
      it("clears the history", () => {
        const initialState = getInitialState({
          history: new Array(10).fill({
            A: 0,
            B: 0,
            id: "",
            value: 0,
            op: "add"
          })
        });
        expect(initialState.history).toHaveLength(10);
        const state = reducer(initialState, actions.historyClear());
        expect(state.history).toHaveLength(0);
      });
    });

    describe("tutorial", () => {
      it("can be set to done", () => {
        const initialState = getInitialState({ tutorialDone: false });
        expect(initialState.tutorialDone).toBeFalsy();
        const state = reducer(initialState, actions.setTutorialDone(true));
        expect(state.tutorialDone).toBeTruthy();
      });

      it("can be set to not done", () => {
        const initialState = getInitialState({ tutorialDone: true });
        expect(initialState.tutorialDone).toBeTruthy();
        const state = reducer(initialState, actions.setTutorialDone(false));
        expect(state.tutorialDone).toBeFalsy();
      });
    });
  });
});
