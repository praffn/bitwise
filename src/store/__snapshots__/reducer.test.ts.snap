// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer initialState has correct initialState 1`] = `
Object {
  "bitLength": 16,
  "history": Array [],
  "radix": "dec",
  "registers": Object {
    "A": 0,
    "B": 0,
  },
  "signed": false,
  "tutorialDone": false,
}
`;
